Server panic on sigint

**Describe the bug**
The waypoint server panic-crashed after sigint via ctrl-c, with the following trace:

```
2021-05-17T15:53:54.024Z [INFO]  waypoint.server.http: HTTP request: POST /hashicorp.waypoint.Waypoint/ListBuilds: date=2021-05-17T15:53:54.023985604Z http.host=34.201.38.155:9702 http.method=POST http.request_path=/hashicorp.waypoint.Waypoint/ListBuilds http.remote_addr=47.6.36.181 http.response_size=2504 http.scheme= http.status_code=200 http.useragent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36" http.version=HTTP/1.1
2021/05/17 15:53:54 http: TLS handshake error from 47.6.36.181:55081: remote error: tls: unknown certificate
2021-05-17T15:53:54.132Z [INFO]  waypoint.server.grpc: /hashicorp.waypoint.Waypoint/ListDeployments response: error=<nil> duration=169.069908ms
2021-05-17T15:53:54.132Z [INFO]  waypoint.server.http: HTTP request: POST /hashicorp.waypoint.Waypoint/ListDeployments: date=2021-05-17T15:53:53.962851867Z http.host=34.201.38.155:9702 http.method=POST http.request_path=/hashicorp.waypoint.Waypoint/ListDeployments http.remote_addr=47.6.36.181 http.response_size=1380 http.scheme= http.status_code=200 http.useragent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36" http.version=HTTP/1.1
^C2021-05-17T15:54:01.869Z [WARN]  waypoint: interrupt received, cancelling context
2021-05-17T15:54:01.870Z [INFO]  waypoint.server.grpc: /hashicorp.waypoint.Waypoint/GetLogStream response: error="context canceled" duration=3m6.962165008s
2021-05-17T15:54:01.870Z [INFO]  waypoint.server.http: HTTP request: POST /hashicorp.waypoint.Waypoint/GetLogStream: date=2021-05-17T15:50:54.907719383Z http.host=34.201.38.155:9702 http.method=POST http.request_path=/hashicorp.waypoint.Waypoint/GetLogStream http.remote_addr=47.6.36.181 http.response_size=0 http.scheme= http.status_code=200 http.useragent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36" http.version=HTTP/1.1
2021-05-17T15:54:01.870Z [INFO]  waypoint.server.grpc: shutting down gRPC server
panic: Drain() is not implemented

goroutine 613 [running]:
google.golang.org/grpc/internal/transport.(*serverHandlerTransport).Drain(0xc000f950e0)
	/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.33.1/internal/transport/handler_server.go:440 +0x39
google.golang.org/grpc.(*Server).GracefulStop(0xc000818000)
	/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.33.1/server.go:1683 +0x385
github.com/hashicorp/waypoint/internal/server.grpcInit.func2.1(0xc000e6cae0, 0x3317c78, 0xc0001383f0, 0xc000818000)
	/home/circleci/project/waypoint/internal/server/grpc.go:73 +0x91
created by github.com/hashicorp/waypoint/internal/server.grpcInit.func2
	/home/circleci/project/waypoint/internal/server/grpc.go:70 +0xa5
```

There was no ongoing build, deploy, or release at the time of the panic. 

**Steps to Reproduce**
Unfortunately, I haven't been able to reproduce this. I was running a server manually on ec2, with this command:

```
waypoint server run -accept-tos -advertise-addr=<addr> -listen-grpc=0.0.0.0:9701 -listen-http=0.0.0.0:9702 -db=/opt/waypoint/data.db -advertise-tls-skip-verify -vvv
```

**Expected behavior**
Graceful shutdown

**Waypoint Platform Versions**
Additional version and platform information to help triage the issue if
applicable:

* Waypoint CLI Version: v0.3.2
* Waypoint Server Platform and Version: v0.3.2

--- a/.changelog/1761.txt
+++ b/.changelog/1761.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+server: Fix an intermittent panic when shutting down the server
+```

--- a/go.mod
+++ b/go.mod
@@ -81,7 +81,6 @@ require (
 	github.com/moby/sys/symlink v0.1.0 // indirect
 	github.com/mr-tron/base58 v1.2.0
 	github.com/natefinch/atomic v0.0.0-20200526193002-18c0533a5b09
-	github.com/oklog/run v1.1.0
 	github.com/oklog/ulid v1.3.1
 	github.com/oklog/ulid/v2 v2.0.2
 	github.com/olekukonko/tablewriter v0.0.4

--- a/internal/server/grpc.go
+++ b/internal/server/grpc.go
@@ -97,19 +97,19 @@ func (s *grpcServer) close() {
 	gracefulCh := make(chan struct{})
 	go func() {
 		defer close(gracefulCh)
-		log.Info("shutting down gRPC server")
+		log.Debug("stopping")
 		s.server.GracefulStop()
 	}()
 
 	select {
 	case <-gracefulCh:
-		log.Debug("grpc server exited gracefully")
+		log.Debug("exited gracefully")
 
 	// After a timeout we just forcibly exit. Our gRPC endpoints should
 	// be fairly quick and their operations are atomic so we just kill
 	// the connections after a few seconds.
 	case <-time.After(2 * time.Second):
-		log.Debug("stopping gRPC server after waiting unsuccessfully for graceful shutdown")
+		log.Debug("stopping forcefully after waiting unsuccessfully for graceful stop")
 		s.server.Stop()
 	}
 }

--- a/internal/server/grpc.go
+++ b/internal/server/grpc.go
@@ -103,6 +103,7 @@ func (s *grpcServer) close() {
 
 	select {
 	case <-gracefulCh:
+		log.Debug("grpc server exited gracefully")
 
 	// After a timeout we just forcibly exit. Our gRPC endpoints should
 	// be fairly quick and their operations are atomic so we just kill

--- a/internal/server/grpc.go
+++ b/internal/server/grpc.go
@@ -13,8 +13,8 @@ import (
 )
 
 type grpcServer struct {
-	opts *options
-	log hclog.Logger
+	opts   *options
+	log    hclog.Logger
 	server *grpc.Server
 }
 
@@ -65,9 +65,9 @@ func newGrpcServer(opts *options) (*grpcServer, error) {
 	s := grpc.NewServer(so...)
 
 	return &grpcServer{
-		opts: opts,
+		opts:   opts,
 		server: s,
-		log: log,
+		log:    log,
 	}, nil
 }
 

--- a/internal/server/grpc.go
+++ b/internal/server/grpc.go
@@ -4,7 +4,7 @@ import (
 	"time"
 
 	"github.com/golang/protobuf/ptypes/empty"
-	"github.com/oklog/run"
+	"github.com/hashicorp/go-hclog"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/keepalive"
 	"google.golang.org/grpc/reflection"
@@ -12,14 +12,20 @@ import (
 	pb "github.com/hashicorp/waypoint/internal/server/gen"
 )
 
-// grpcInit initializes the gRPC server and adds it to the run group.
-func grpcInit(group *run.Group, opts *options) error {
+type grpcServer struct {
+	opts *options
+	log hclog.Logger
+	server *grpc.Server
+}
+
+// newGrpcServer initializes a new gRPC server
+func newGrpcServer(opts *options) (*grpcServer, error) {
 	log := opts.Logger.Named("grpc")
 
 	// Get our server info immediately
 	resp, err := opts.Service.GetVersionInfo(opts.Context, &empty.Empty{})
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	var so []grpc.ServerOption
@@ -57,42 +63,53 @@ func grpcInit(group *run.Group, opts *options) error {
 	}
 
 	s := grpc.NewServer(so...)
-	opts.grpcServer = s
 
+	return &grpcServer{
+		opts: opts,
+		server: s,
+		log: log,
+	}, nil
+}
+
+// start starts the grpc server
+func (s *grpcServer) start() error {
 	// Register the reflection service. This makes using tools like grpcurl
 	// easier. It makes it slightly easier for malicious users to know about
 	// the service but I think they'd figure out its a waypoint server
 	// easy enough.
-	reflection.Register(s)
+	reflection.Register(s.server)
 
 	// Register our server
-	pb.RegisterWaypointServer(s, opts.Service)
+	pb.RegisterWaypointServer(s.server, s.opts.Service)
+	// Serve traffic
+	ln := s.opts.GRPCListener
+	s.log.Info("starting gRPC server", "addr", ln.Addr().String())
+	return s.server.Serve(ln)
+}
 
-	// Add our gRPC server to the run group
-	group.Add(func() error {
-		// Serve traffic
-		ln := opts.GRPCListener
-		log.Info("starting gRPC server", "addr", ln.Addr().String())
-		return s.Serve(ln)
-	}, func(err error) {
-		// Graceful in a goroutine so we can timeout
-		gracefulCh := make(chan struct{})
-		go func() {
-			defer close(gracefulCh)
-			log.Info("shutting down gRPC server")
-			s.GracefulStop()
-		}()
+// close stops the grpc server, gracefully if possible.
+// Warning: before closing the GRPC server, the HTTP server must first be closed.
+// Attempting to gracefully stop the GRPC server first will cause it to drain HTTP connections,
+// which will panic.
+func (s *grpcServer) close() {
+	log := s.log
+	// Graceful in a goroutine so we can timeout
+	gracefulCh := make(chan struct{})
+	go func() {
+		defer close(gracefulCh)
+		log.Info("shutting down gRPC server")
+		s.server.GracefulStop()
+	}()
 
-		select {
-		case <-gracefulCh:
+	select {
+	case <-gracefulCh:
 
-		// After a timeout we just forcibly exit. Our gRPC endpoints should
-		// be fairly quick and their operations are atomic so we just kill
-		// the connections after a few seconds.
-		case <-time.After(2 * time.Second):
-			s.Stop()
-		}
-	})
-
-	return nil
+	// After a timeout we just forcibly exit. Our gRPC endpoints should
+	// be fairly quick and their operations are atomic so we just kill
+	// the connections after a few seconds.
+	case <-time.After(2 * time.Second):
+		log.Debug("stopping gRPC server after waiting unsuccessfully for graceful shutdown")
+		s.server.Stop()
+	}
+	return
 }

--- a/internal/server/grpc.go
+++ b/internal/server/grpc.go
@@ -111,5 +111,4 @@ func (s *grpcServer) close() {
 		log.Debug("stopping gRPC server after waiting unsuccessfully for graceful shutdown")
 		s.server.Stop()
 	}
-	return
 }

--- a/internal/server/http.go
+++ b/internal/server/http.go
@@ -90,20 +90,21 @@ func (s *httpServer) close() {
 	gracefulCh := make(chan struct{})
 	go func() {
 		defer close(gracefulCh)
-		log.Info("shutting down HTTP server")
+		log.Info("stopping")
 		if err := s.server.Shutdown(ctx); err != nil {
-			log.Error("failed gracefully shutting down http server: %s", err)
+			log.Error("failed graceful shutdown: %s", err)
 		}
 	}()
 
 	select {
 	case <-gracefulCh:
+		log.Debug("exited gracefully")
 
 	// After a timeout we just forcibly exit. Our HTTP endpoints should
 	// be fairly quick and their operations are atomic so we just kill
 	// the connections after a few seconds.
 	case <-time.After(2 * time.Second):
-		log.Debug("stopping http server after waiting unsuccessfully for graceful shutdown")
+		log.Debug("stopping forcefully after waiting unsuccessfully for graceful stop")
 		cancelFunc()
 	}
 }

--- a/internal/server/http.go
+++ b/internal/server/http.go
@@ -2,6 +2,8 @@ package server
 
 import (
 	"context"
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/grpc"
 	"net"
 	"net/http"
 	"strings"
@@ -10,11 +12,17 @@ import (
 	assetfs "github.com/elazarl/go-bindata-assetfs"
 	"github.com/hashicorp/waypoint/internal/server/gen"
 	"github.com/improbable-eng/grpc-web/go/grpcweb"
-	"github.com/oklog/run"
 )
 
-// httpInit initializes the HTTP server and adds it to the run group.
-func httpInit(group *run.Group, opts *options) error {
+type httpServer struct {
+	opts *options
+	log hclog.Logger
+	server *http.Server
+}
+
+// newHttpServer initializes a new http server.
+// Uses grpc-web to wrap an existing grpc server.
+func newHttpServer(grpcServer *grpc.Server, opts *options) *httpServer{
 	log := opts.Logger.Named("http")
 	if opts.HTTPListener == nil {
 		log.Info("HTTP listener not specified, HTTP API is disabled")
@@ -22,7 +30,7 @@ func httpInit(group *run.Group, opts *options) error {
 	}
 
 	// Wrap the grpc server so that it is grpc-web compatible
-	grpcWrapped := grpcweb.WrapServer(opts.grpcServer,
+	grpcWrapped := grpcweb.WrapServer(grpcServer,
 		grpcweb.WithCorsForRegisteredEndpointsOnly(false),
 		grpcweb.WithOriginFunc(func(string) bool { return true }),
 		grpcweb.WithAllowNonRootResource(true),
@@ -47,43 +55,55 @@ func httpInit(group *run.Group, opts *options) error {
 	})
 
 	// Create our http server
-	httpSrv := &http.Server{
-		ReadHeaderTimeout: 5 * time.Second,
-		IdleTimeout:       120 * time.Second,
-		Handler:           httpLogHandler(rootHandler, log),
-		BaseContext: func(net.Listener) context.Context {
-			return opts.Context
+	return &httpServer{
+		opts: opts,
+		log: log,
+		server: &http.Server{
+			ReadHeaderTimeout: 5 * time.Second,
+			IdleTimeout:       120 * time.Second,
+			Handler:           httpLogHandler(rootHandler, log),
+			BaseContext: func(net.Listener) context.Context {
+				return opts.Context
+			},
 		},
 	}
+}
 
-	// Add our gRPC server to the run group
-	group.Add(func() error {
-		// Serve traffic
-		ln := opts.HTTPListener
-		log.Info("starting HTTP server", "addr", ln.Addr().String())
-		return httpSrv.Serve(ln)
-	}, func(err error) {
-		ctx, cancelFunc := context.WithCancel(context.Background())
-		defer cancelFunc()
+// start starts an http server
+func (s *httpServer) start() error {
+	// Serve traffic
+	ln := s.opts.HTTPListener
+	s.log.Info("starting HTTP server", "addr", ln.Addr().String())
+	return s.server.Serve(ln)
+}
 
-		// Graceful in a goroutine so we can timeout
-		gracefulCh := make(chan struct{})
-		go func() {
-			defer close(gracefulCh)
-			log.Info("shutting down HTTP server")
-			httpSrv.Shutdown(ctx)
-		}()
+// close stops the grpc server, gracefully if possible.
+// Warning: before closing the GRPC server, this HTTP server must first be closed.
+// Attempting to gracefully stop the GRPC server first will cause it to drain HTTP connections,
+// which will panic.
+func (s *httpServer) close() {
+	log := s.log
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
 
-		select {
-		case <-gracefulCh:
-
-		// After a timeout we just forcibly exit. Our HTTP endpoints should
-		// be fairly quick and their operations are atomic so we just kill
-		// the connections after a few seconds.
-		case <-time.After(2 * time.Second):
-			cancelFunc()
+	// Graceful in a goroutine so we can timeout
+	gracefulCh := make(chan struct{})
+	go func() {
+		defer close(gracefulCh)
+		log.Info("shutting down HTTP server")
+		if err := s.server.Shutdown(ctx); err != nil {
+			log.Error("failed gracefully shutting down http server: %s", err)
 		}
-	})
+	}()
 
-	return nil
+	select {
+	case <-gracefulCh:
+
+	// After a timeout we just forcibly exit. Our HTTP endpoints should
+	// be fairly quick and their operations are atomic so we just kill
+	// the connections after a few seconds.
+	case <-time.After(2 * time.Second):
+		log.Debug("stopping http server after waiting unsuccessfully for graceful shutdown")
+		cancelFunc()
+	}
 }

--- a/internal/server/http.go
+++ b/internal/server/http.go
@@ -15,14 +15,14 @@ import (
 )
 
 type httpServer struct {
-	opts *options
-	log hclog.Logger
+	opts   *options
+	log    hclog.Logger
 	server *http.Server
 }
 
 // newHttpServer initializes a new http server.
 // Uses grpc-web to wrap an existing grpc server.
-func newHttpServer(grpcServer *grpc.Server, opts *options) *httpServer{
+func newHttpServer(grpcServer *grpc.Server, opts *options) *httpServer {
 	log := opts.Logger.Named("http")
 	if opts.HTTPListener == nil {
 		log.Info("HTTP listener not specified, HTTP API is disabled")
@@ -57,7 +57,7 @@ func newHttpServer(grpcServer *grpc.Server, opts *options) *httpServer{
 	// Create our http server
 	return &httpServer{
 		opts: opts,
-		log: log,
+		log:  log,
 		server: &http.Server{
 			ReadHeaderTimeout: 5 * time.Second,
 			IdleTimeout:       120 * time.Second,

--- a/internal/server/http.go
+++ b/internal/server/http.go
@@ -77,7 +77,7 @@ func (s *httpServer) start() error {
 	return s.server.Serve(ln)
 }
 
-// close stops the grpc server, gracefully if possible.
+// close stops the grpc server, gracefully if possible. Should be called exactly once.
 // Warning: before closing the GRPC server, this HTTP server must first be closed.
 // Attempting to gracefully stop the GRPC server first will cause it to drain HTTP connections,
 // which will panic.
@@ -90,7 +90,7 @@ func (s *httpServer) close() {
 	gracefulCh := make(chan struct{})
 	go func() {
 		defer close(gracefulCh)
-		log.Info("stopping")
+		log.Debug("stopping")
 		if err := s.server.Shutdown(ctx); err != nil {
 			log.Error("failed graceful shutdown: %s", err)
 		}

--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -52,9 +52,9 @@ func Run(opts ...Option) error {
 	defer cancel()
 
 	select {
-	case err := <- errch:
+	case err := <-errch:
 		return err
-	case <- cfg.Context.Done():
+	case <-cfg.Context.Done():
 		// Must shut down the http server first, as the grpc server can't drain http connections
 		httpServer.close()
 		grpcServer.close()

--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -34,7 +34,7 @@ func Run(opts ...Option) error {
 		return err
 	}
 
-	errch := make(chan error, 0)
+	errch := make(chan error)
 	go func() {
 		if err := grpcServer.start(); err != nil {
 			errch <- err

--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -2,7 +2,9 @@ package server
 
 import (
 	"context"
+	"fmt"
 	"net"
+	"sync"
 
 	"github.com/hashicorp/go-hclog"
 	pb "github.com/hashicorp/waypoint/internal/server/gen"
@@ -29,38 +31,66 @@ func Run(opts ...Option) error {
 		cfg.Logger = hclog.L()
 	}
 
+	log := cfg.Logger.Named("server")
+
+	wg := sync.WaitGroup{}
+
 	grpcServer, err := newGrpcServer(&cfg)
 	if err != nil {
 		return err
 	}
 
-	errch := make(chan error)
+	gprcErrs := make(chan error, 1)
+	wg.Add(1)
 	go func() {
 		if err := grpcServer.start(); err != nil {
-			errch <- err
+			log.Debug("GRPC server has exited", "err", err)
+			gprcErrs <- err
 		}
+		wg.Done()
+		log.Debug("GRPC server has exited with no err")
 	}()
 
+	httpErrs := make(chan error, 1)
 	httpEnabled := cfg.HTTPListener != nil
 	var httpServer *httpServer
 	if httpEnabled {
+		wg.Add(1)
 		httpServer = newHttpServer(grpcServer.server, &cfg)
 		go func() {
 			if err := httpServer.start(); err != nil {
-				errch <- err
+				log.Debug("HTTP server has exited", "err", err)
+				httpErrs <- err
 			}
+			wg.Done()
+			log.Debug("HTTP server has exited with no err")
 		}()
 	}
 
 	ctx, cancel := context.WithCancel(cfg.Context)
 	defer cancel()
 
+	defer func() {
+		log.Debug("Waiting for all server run processes to end")
+		wg.Wait()
+		log.Debug("Server processes have exited.")
+	}()
+
 	select {
-	case err := <-errch:
-		return err
-	case <-cfg.Context.Done():
+	case err := <-gprcErrs:
+		// If the GRPC server errored, we can assume it's closed and shut down the http server if necessary
 		if httpEnabled {
-			// Must shut down the http server first, as the grpc server can't drain http connections
+			httpServer.close()
+		}
+		return fmt.Errorf("failed running the grpc server: %w", err)
+	case err := <-httpErrs:
+		// If the HTTP server errored, we can assume it's closed and shut down the grpc server
+		grpcServer.close()
+		return fmt.Errorf("failed running the http server: %w", err)
+	case <-cfg.Context.Done():
+		// Received an external shutdown signal, and should close everything.
+		// NOTE: must close HTTP server before GRPC server.
+		if httpEnabled {
 			httpServer.close()
 		}
 		grpcServer.close()

--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -41,12 +41,16 @@ func Run(opts ...Option) error {
 		}
 	}()
 
-	httpServer := newHttpServer(grpcServer.server, &cfg)
-	go func() {
-		if err := httpServer.start(); err != nil {
-			errch <- err
-		}
-	}()
+	httpEnabled := cfg.HTTPListener != nil
+	var httpServer *httpServer
+	if httpEnabled {
+		httpServer = newHttpServer(grpcServer.server, &cfg)
+		go func() {
+			if err := httpServer.start(); err != nil {
+				errch <- err
+			}
+		}()
+	}
 
 	ctx, cancel := context.WithCancel(cfg.Context)
 	defer cancel()
@@ -55,8 +59,10 @@ func Run(opts ...Option) error {
 	case err := <-errch:
 		return err
 	case <-cfg.Context.Done():
-		// Must shut down the http server first, as the grpc server can't drain http connections
-		httpServer.close()
+		if httpEnabled {
+			// Must shut down the http server first, as the grpc server can't drain http connections
+			httpServer.close()
+		}
 		grpcServer.close()
 		return ctx.Err()
 	}

--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -2,7 +2,6 @@ package server
 
 import (
 	"context"
-	"fmt"
 	"net"
 	"sync"
 
@@ -28,10 +27,9 @@ func Run(opts ...Option) error {
 		cfg.Context = context.Background()
 	}
 	if cfg.Logger == nil {
-		cfg.Logger = hclog.L()
+		cfg.Logger = hclog.L().Named("server")
 	}
-
-	log := cfg.Logger.Named("server")
+	log := cfg.Logger
 
 	grpcServer, err := newGrpcServer(&cfg)
 	if err != nil {
@@ -73,11 +71,13 @@ func Run(opts ...Option) error {
 		if httpEnabled {
 			httpServer.close()
 		}
-		return fmt.Errorf("failed running the grpc server: %w", err)
+		log.Error("failed running the grpc server", "err", err)
+		return err
 	case err := <-httpErrs:
 		// If the HTTP server errored, we can assume it's closed and shut down the grpc server
 		grpcServer.close()
-		return fmt.Errorf("failed running the http server: %w", err)
+		log.Error("failed running the http server", "err", err)
+		return err
 	case <-cfg.Context.Done():
 		// Received an external shutdown signal, and should close everything.
 		// NOTE: must close HTTP server before GRPC server.

--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -5,9 +5,6 @@ import (
 	"net"
 
 	"github.com/hashicorp/go-hclog"
-	"github.com/oklog/run"
-	"google.golang.org/grpc"
-
 	pb "github.com/hashicorp/waypoint/internal/server/gen"
 )
 
@@ -32,32 +29,37 @@ func Run(opts ...Option) error {
 		cfg.Logger = hclog.L()
 	}
 
-	// Setup our run group since we're going to be starting multiple
-	// goroutines for all the servers that we want to live/die as a group.
-	var group run.Group
+	grpcServer, err := newGrpcServer(&cfg)
+	if err != nil {
+		return err
+	}
 
-	// We first add an actor that just returns when the context ends. This
-	// will trigger the rest of the group to end since a group will not exit
-	// until any of its actors exit.
-	ctx, cancelCtx := context.WithCancel(cfg.Context)
-	cfg.Context = ctx
-	group.Add(func() error {
-		<-ctx.Done()
+	errch := make(chan error, 0)
+	go func() {
+		if err := grpcServer.start(); err != nil {
+			errch <- err
+		}
+	}()
+
+	httpServer := newHttpServer(grpcServer.server, &cfg)
+	go func() {
+		if err := httpServer.start(); err != nil {
+			errch <- err
+		}
+	}()
+
+	ctx, cancel := context.WithCancel(cfg.Context)
+	defer cancel()
+
+	select {
+	case err := <- errch:
+		return err
+	case <- cfg.Context.Done():
+		// Must shut down the http server first, as the grpc server can't drain http connections
+		httpServer.close()
+		grpcServer.close()
 		return ctx.Err()
-	}, func(error) { cancelCtx() })
-
-	// Setup our gRPC server.
-	if err := grpcInit(&group, &cfg); err != nil {
-		return err
 	}
-
-	// Setup our HTTP server.
-	if err := httpInit(&group, &cfg); err != nil {
-		return err
-	}
-
-	// Run!
-	return group.Run()
 }
 
 // Option configures Run
@@ -90,8 +92,6 @@ type options struct {
 
 	// BrowserUIEnabled determines if the browser UI should be mounted
 	BrowserUIEnabled bool
-
-	grpcServer *grpc.Server
 }
 
 // WithContext sets the context for the server. When this context is cancelled,

--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -33,22 +33,20 @@ func Run(opts ...Option) error {
 
 	log := cfg.Logger.Named("server")
 
-	wg := sync.WaitGroup{}
-
 	grpcServer, err := newGrpcServer(&cfg)
 	if err != nil {
 		return err
 	}
 
+	wg := sync.WaitGroup{}
+
 	gprcErrs := make(chan error, 1)
 	wg.Add(1)
 	go func() {
-		if err := grpcServer.start(); err != nil {
-			log.Debug("GRPC server has exited", "err", err)
-			gprcErrs <- err
-		}
+		err := grpcServer.start()
+		gprcErrs <- err
+		log.Debug("gRPC server has exited", "err", err)
 		wg.Done()
-		log.Debug("GRPC server has exited with no err")
 	}()
 
 	httpErrs := make(chan error, 1)
@@ -58,24 +56,17 @@ func Run(opts ...Option) error {
 		wg.Add(1)
 		httpServer = newHttpServer(grpcServer.server, &cfg)
 		go func() {
-			if err := httpServer.start(); err != nil {
-				log.Debug("HTTP server has exited", "err", err)
-				httpErrs <- err
-			}
+			err := httpServer.start()
+			httpErrs <- err
+			log.Debug("HTTP server has exited", "err", err)
 			wg.Done()
-			log.Debug("HTTP server has exited with no err")
 		}()
 	}
 
 	ctx, cancel := context.WithCancel(cfg.Context)
 	defer cancel()
 
-	defer func() {
-		log.Debug("Waiting for all server run processes to end")
-		wg.Wait()
-		log.Debug("Server processes have exited.")
-	}()
-
+	defer wg.Wait() // Wait for server run processes to exit before returning
 	select {
 	case err := <-gprcErrs:
 		// If the GRPC server errored, we can assume it's closed and shut down the http server if necessary

